code/Model/SiteTree.php

```diff
@@ -1614,36 +1614,27 @@ public function validate()
      */
     public function validURLSegment()
     {
-        $excludes = Director::config()->get('rules');
-        $excludes = array_keys($excludes);
-        $disallowedSegments = array_map(function ($key) {
-            $route = explode('/', $key);
-            if (!empty($route) && strpos($route[0], '$') === false) {
-                return $route[0];
+        // Check known urlsegment blacklists
+        if (!self::config()->nested_urls || !$this->ParentID) {
+            // Guard against url segments for the base page
+            if (in_array(strtolower($this->URLSegment), $this->getExcludedURLSegments())) {
+                // Default to '-2', onBeforeWrite takes care of further possible clashes
+                return false;
             }
-            return;
-        }, $excludes);
-
-        if (!$this->ParentID && in_array($this->URLSegment, $disallowedSegments)) {
-            // Default to '-2', onBeforeWrite takes care of further possible clashes
-            return false;
-        }
-
-
-        if (self::config()->nested_urls && $parent = $this->Parent()) {
+            // Check if urlsegment matches a class name
+            if (class_exists($this->URLSegment) && is_subclass_of($this->URLSegment, RequestHandler::class)) {
```

Comment:
I thought Controller urlsegment access was removed during alpha?

Reply:
Yes but this seems a prudent check in any case.

I was worried if I removed it, it would be a risk.